Added ability to report spelling errors in braille

nvdaes opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests.

Please initially open PRs as a draft.
When you would like a review, mark the PR as "ready for review".
See https://github.com/nvaccess/nvda/blob/master/.github/CONTRIBUTING.md.
-->

### Link to issue number:
<!-- Use Closes/Fixes/Resolves #xxx to link this PR to the issue it is responding to. -->
Fixes #7608
### Summary of the issue:
Currently, spelling errors aren't presented in braille. This may negatively impact deaf-blind people and other braille users.
### Description of user facing changes:
If NVDA is configured to report format changes with tags in braille, and to report spelling errors in braille, these mistakes would be presented in braille with the "e" and "e with dot 7" tag.
An unassigned command has been added to toggle the reporting of spelling errors in braille. This can also be configured from the Document Formatting dialog.
### Description of developer facing changes:

None.
### Description of development approach:
Added a new item in the `fontAttributeFormattingMarkers` dictionary of `braille.py` for spelling errors (`invalid-spelling`).
A `shouldBeUsed` method has been added to the `FormattingMarker`class, to determine if an attribute should be reported.
### Testing strategy:
Tested locally in Microsoft Word and Notepad, typing text detected as spelling errors.
### Known issues with pull request:
None.
### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

<!-- Please keep the following -->
@coderabbitai summary

nvdaes committed:
Added ability to report spelling errors in braille

josephsl commented:
Hi, interesting. May I suggest using a Unicode braille pattern to surround the spelling/grammar error so people would not become confused, thinking that “m” is part of the text when it isn’t? Thanks.

SaschaCowley commented:
@nvdaes I agree with @CyrilleB79 that this should not be controlled by the "font attributes" option. I suggest that we change the "spelling errors" option to allow the user to select speech and/or braille as can be done with font attributes. This will require changing the way NVDA calculates the indicators in braille. I think the approach I would take would be to add a callable to `FormattingMarker` that returns `True` if the marker should be used if appropriate, and `False` if not. We can then refactor the code to use that callable rather than the overarching check as done currently. What do you think?

nvdaes committed:
Add a callable to decide if formatting markers should be added in braille

nvdaes committed:
Add a key param to shouldBeUsed method

nvdaes committed:
Use key param

CyrilleB79 commented:
Why did you use "m" letter while we have "Spelling errors" in the UI, and not "Mistakes"? Have you considered using "e" instead?
It's worth noting that "s" cannot be used since it is already used by "strikethrough"

nvdaes commented:
> Why did you use "m" letter while we have "Spelling errors" in the UI, and not "Mistakes"? Have you considered using "e" instead? It's worth noting that "s" cannot be used since it is already used by "strikethrough"


Yes, in fact I have considered using "e". I'll change this and will continue my work when your PR to add ability to use audio is merged.
I think we can use a single line in the changelog mentioning our names and the corresponding issues, something like:
* Added ability to report spelling errors with audio and braille (@CyrilleB79, @nvdaes)

CyrilleB79 commented:
@SaschaCowley you wrote:
> @nvdaes I agree with @CyrilleB79 that this should not be controlled by the "font attributes" option. I suggest that we change the "spelling errors" option to allow the user to select speech and/or braille as can be done with font attributes. 

What do you suggest regarding the GUI and taking into account the work in #17997 (see more specifically https://github.com/nvaccess/nvda/pull/17997#issuecomment-3169232755)?

@nvdaes you wrote:

> I think we can use a single line in the changelog mentioning our names and the corresponding issues, something like:
> 
>     * Added ability to report spelling errors with audio and braille (@CyrilleB79, @nvdaes)

Feel free to do so when the final design is clearer and if you still find it appropriate.

SaschaCowley commented:
@CyrilleB79  @nvdaes We think that a check list with "speech", "sound" and "braille" options is most appropriate in this case. These could be internally stored as a bit flag.

nvdaes commented:
Sascha wrote:

> We think that a check list with "speech", "sound" and "braille" options is most appropriate in this case. These could be internally stored as a bit flag.

I agree with this proposal. I'll continue my work, as mentioned, when Cyrille's work about sound and speech is merged.

CyrilleB79 commented:
How would you modify the toggle script?

seanbudd commented:
Sorry wrong PR

nvdaes committed:
Merge remote-tracking branch 'nvaccess/master' into brailleErrors

nvdaes committed:
Add braille to ReportSpellingErrors

nvdaes committed:
Add ability to report spelling errors in braille, and set default to speech and braille

nvdaes committed:
Add script to toggle reporting spelling errors in braille

pre-commit-ci[bot] committed:
Pre-commit auto-fix

nvdaes commented:
I think we should make the following modes available to cycle with a script to report spelling errors:

- Off.
- Sounds.
- Speech and sounds.
And, additionally, a different script can be used to turn on and off reporting spelling errors in braille.

Probably the toggle integer value is not appropriate here.

nvdaes committed:
Improve script to cycle through audio feedback for spelling errors

nvdaes committed:
Update user guide

nvdaes committed:
Update user guide

nvdaes commented:
@CyrilleB79 , think that I've addressed your review.
I see that the toggleIntegerValue function is not documented in changes.md, but the toggle boolean function was documented, I haven't added a note about the toggleBinarieValue function, but perhaps it would be useful.
Also, to turn on and off single values of int flags like Reporting errors in braille ond or off, I used the tilde and it's not easy to find, at least not in all keyboards. Perhaps a function for this may be added.
If you don't have other comments, I'll request a review from NV Access soon, as this works for me.

CyrilleB79 reviewed (commented):
Thanks for integrating the latest changes in master.

I have not tested with a braille display, but have tried all the config stuff.

nvdaes committed:
Apply suggestions from code review

Co-authored-by: Cyrille Bougot <cyrille.bougot2@laposte.net>

nvdaes committed:
Address review fixing script to toggle audio options to report spelling errors

pre-commit-ci[bot] committed:
Pre-commit auto-fix

nvdaes committed:
Address review

CyrilleB79 commented on source/globalCommands.py:
```diff
@@ -793,15 +824,40 @@ def script_toggleReportStyle(self, gesture):
 		category=SCRCAT_DOCUMENTFORMATTING,
 	)
 	def script_toggleReportSpellingErrors(self, gesture: inputCore.InputGesture):
```
I have not tested this. But reading it, the code seems to be working as expected.

I feel that doing such specific cases with numeric values is not so self explaining.
Why not something like the following:
```
brailleValue = config.conf["documentFormatting"]["reportSpellingErrors2"] & ReportSpellingErrors.BRAILLE
speechAndSoundValue = config.conf["documentFormatting"]["reportSpellingErrors2"] & (ReportSpellingErrors.SPEECH | ReportSpellingErrors.SOUND)
speechAndSoundValue = (speechAndSoundValue + 1) % 4  # Or may we replace 4 by a more self explaining value?
config.conf["documentFormatting"]["reportSpellingErrors2"] = brailleValue | speechAndSoundValue
ui.message(

			_("Report spelling errors {mode}").format(mode=ReportSpellingErrors(speechAndSoundValue).displayString)

		)
```

nvdaes committed:
Merge branch 'brailleErrors' of https://github.com/nvdaes/nvda into brailleErrors

nvdaes committed:
Merge branch 'master' into brailleErrors

pre-commit-ci[bot] committed:
Pre-commit auto-fix

nvdaes committed:
Improve script

nvdaes committed:
Merge remote-tracking branch 'nvaccess/master' into brailleErrors

nvdaes committed:
Merge branch 'master' into brailleErrors

nvdaes committed:
Merge remote-tracking branch 'origin/master' into brailleErrors

nvdaes committed:
Merge remote-tracking branch 'origin/master' into brailleErrors

nvdaes committed:
Merge branch 'master' into brailleErrors

nvdaes committed:
Merge remote-tracking branch 'origin/master' into brailleErrors

nvdaes committed:
Fix typo in user guide anchor

nvdaes committed:
Added changes for developers

SaschaCowley reviewed (changes requested):
Looks good, just a few suggestions.

nvdaes committed:
Apply suggestions from code review

Co-authored-by: Sascha Cowley <16543535+SaschaCowley@users.noreply.github.com>

nvdaes committed:
Address review: remove unused function

nvdaes committed:
Address review: simplify script, needs to be tested

nvdaes committed:
Apply review suggestion, to be tested later

nvdaes commented:
Thanks a lot @SaschaCowley ! I love these simplifications in the code. I'll test this later just to be asbolutely sure, and I'll provide feedback here.

nvdaes committed:
Fix script

nvdaes committed:
Merge remote-tracking branch 'origin/master' into brailleErrors

pre-commit-ci[bot] committed:
Pre-commit auto-fix

nvdaes commented:
@SaschaCowley , thanks for your suggestions. I think that this is ready for review.

Qchristensen reviewed (approved):
Reads well.
